src/frontend/profile_utils.rs

```diff
@@ -0,0 +1,34 @@
+use std::time::{Duration, Instant};
+
+// Simple timer intended for profiling
+// Lazily inits. Returns `Duration`s or raw string (in ms)
+// Use Option as a safeguard because apparently previous Rust versions panicked
+// when current time was earlier than self.
+
+#[derive(Default)]
+pub struct Timer {
+    start_time: Option<Instant>,
+}
+
+impl Timer {
+    pub fn new() -> Self {
+        Self { start_time: None }
+    }
+
+    pub fn start_timer(&mut self) {
+        self.start_time = Some(Instant::now())
+    }
+
+    pub fn elapsed(&self) -> Option<Duration> {
+        self.start_time.map(|start| start.elapsed())
+    }
```

Comment:
I'm not a Rust professional, so take this with a grain of salt.

It would be safer for the `start_time` field to be an `Instant`, without the `Option` wrapping. The timers added in this PR are started right after they are constructed. Thus, I don't see a point in splitting the `start_timer` method from the `new` constructor. We could avoid the unsafe `unwrap` at callsites of `elapsed` if we avoided `Option`s for `start_time`. If there is a timer that may not have started yet, we can use `Option<Timer>` rather than `Timer`+`start_time: Option<_>`

```suggestion
#[derive(Default)]
pub struct Timer {
    start_time: Instant,
}

impl Timer {
    pub fn new() -> Self {
        Self { start_time: Instant::now() }
    }

    pub fn elapsed(&self) -> Duration {
        self.start_time.elapsed()
    }
```

Reply:
Agreed; also at this point it is questionable whether the `Timer` struct has any purpose, or should we skip the indirection and just use `Instant` directly in `uncached_read_write_query`/`cached_read_query`.

Reply:
> Do you feel strongly I should drop Timer and use Instant directly?

If there's a need to accommodate some additional metrics (e.g. the ones from the plan) we can some fitting encapsulation later. For now it seems simpler to go with the bare `Instant`.

Reply:
Done, thanks!